src/meta/MetaHttpIngestHandler.cpp

```diff
@@ -0,0 +1,146 @@
+    kvstore_ = kvstore;
+    CHECK_NOTNULL(kvstore_);
+}
+
+void MetaHttpIngestHandler::onRequest(std::unique_ptr<HTTPMessage> headers) noexcept {
+    if (headers->getMethod().value() != HTTPMethod::GET) {
+        // Unsupported method
+        err_ = HttpCode::E_UNSUPPORTED_METHOD;
+        return;
+    }
+
+    if (!headers->hasQueryParam("path") ||
+        !headers->hasQueryParam("space")) {
+        err_ = HttpCode::E_ILLEGAL_ARGUMENT;
+        return;
+    }
+
+    path_ = headers->getQueryParam("path");
+    space_ = headers->getIntQueryParam("space");
+}
+
+void MetaHttpIngestHandler::onBody(std::unique_ptr<folly::IOBuf>) noexcept {
+    // Do nothing, we only support GET
+}
+
+
+void MetaHttpIngestHandler::onEOM() noexcept {
+    switch (err_) {
+        case HttpCode::E_UNSUPPORTED_METHOD:
+            ResponseBuilder(downstream_)
+                .status(405, "Method Not Allowed")
+                .sendWithEOM();
+            return;
+        case HttpCode::E_ILLEGAL_ARGUMENT:
+            ResponseBuilder(downstream_)
+                .status(400, "Bad Request")
+                .sendWithEOM();
+            return;
+        default:
+            break;
+    }
+
+    if (ingestSSTFiles(space_, path_)) {
+        LOG(INFO) << "SSTFile ingest successfully " << path_;
+        ResponseBuilder(downstream_)
+            .status(200, "SSTFile ingest successfully")
+            .body("SSTFile ingest successfully")
+            .sendWithEOM();
+    } else {
+        LOG(ERROR) << "SSTFile ingest failed";
+        ResponseBuilder(downstream_)
+            .status(404, "SSTFile ingest failed")
+            .body("SSTFile ingest failed")
+            .sendWithEOM();
+    }
+}
+
+void MetaHttpIngestHandler::onUpgrade(UpgradeProtocol) noexcept {
+    // Do nothing
+}
+
+
+void MetaHttpIngestHandler::requestComplete() noexcept {
+    delete this;
+}
+
+
+void MetaHttpIngestHandler::onError(ProxygenError error) noexcept {
+    LOG(ERROR) << "Web Service MetaHttpIngestHandler got error : "
+               << proxygen::getErrorString(error);
+}
+
+bool MetaHttpIngestHandler::ingestSSTFiles(GraphSpaceID space, const std::string& path) {
+    std::unique_ptr<kvstore::KVIterator> iter;
+    auto prefix = MetaServiceUtils::partPrefix(space);
+    auto ret = kvstore_->prefix(0, 0, prefix, &iter);
+    if (ret != kvstore::ResultCode::SUCCEEDED) {
+        LOG(ERROR) << "Fetch Parts Failed";
+        return false;
+    }
+
+    std::vector<std::string> storageIPs;
+    while (iter->valid()) {
+        for (auto host : MetaServiceUtils::parsePartVal(iter->val())) {
+            auto storageIP = network::NetworkUtils::intToIPv4(host.get_ip());
+            if (std::find(storageIPs.begin(), storageIPs.end(), storageIP) == storageIPs.end()) {
+                storageIPs.push_back(std::move(storageIP));
+            }
+        }
+        iter->next();
+    }
+
+    std::atomic<int> completed(0);
+    std::vector<std::thread> threads;
```

Comment:
I think it is more friendly that use GenericThreadPool to replace threads, if the cluster has more than 1000’s server, this will waste too many cpu resource, and more dangle is that meta server is center of our cluster.